KVM: handle planar modes using MMIO region instead of page faults

bartoldeman opened the pull request:
So now the only remaining in-KVM page protections are for DPMI, where it's necessary, and the low 1MB is all treated like physical RAM, for future compatibility with VCPI (and perhaps real mode as well).

The KVM MMIO interface is a little complex to exit from, as one needs to re-enter KVM to finish the operation, and sometimes twice, for read-modify-write.

With older kernels it's even more complex, as immediate exits from KVM need to be programmed using a signal (with a dummy signal handler that is never actually called) that is masked outside KVM but unmasked inside KVM_RUN.

bartoldeman committed:
vgaemu: adjust instremu in one place for consistency

vgaemu_adjust_instremu always adjusts instremu and also makes all pages
(dirty and clean) PROT_NONE

bartoldeman committed:
KVM/vgaemu: use MMIO region for planar i/o.

Note that when exiting for MMIO you must return to KVM immediately for a
clean exit. The clean exit can either be done using a signal that is only
unblocked during KVM_RUN, or more simply, on newer kernels using
run->immediate_exit=1. This patch only implements the newer method.

stsp commented:
So `vgaemu_adjust_instremu()` appears to
be needed after all?

stsp commented on src/base/emu-i386/kvm.c:
```diff
@@ -1021,6 +1080,59 @@ static unsigned int kvm_run(void)
+	run->immediate_exit = 1;
+      else
+#endif
+	pthread_kill(pthread_self(), KVM_IMMEDIATE_EXIT_SIG);
```
Can we not have that? :)
If immediate_exit works, why to care?
At least please make it a separate patch,
for an easy revert in the future.

bartoldeman replied:
Yes, good idea, will do that (the separate patch).

It's needed on kernels 4.12 and older (think enterprise distros such as RHEL7, everything else has something newer).

stsp replied:
Maybe under !KVM_CAP_IMMEDIATE_EXIT
at least?

bartoldeman replied:
I put almost all the compat code together and in a separate patch now.

bartoldeman commented:
> So `vgaemu_adjust_instremu()` appears to be needed after all?

I decided to play it safe here and not do a profound analysis there is some double work going on but it doesn't hurt.
I did change it to set the prot to NONE on all pages, not just dirty ones, to avoid r/o protected pages.

bartoldeman committed:
KVM: implement compat immediate exit method for older kernels (4.12 and below)

The code is only compiled for
!defined(DISABLE_SYSTEM_WA) || !defined(KVM_CAP_IMMEDIATE_EXIT)
and this patch can easily be reverted once 4.12 is completely obsolete.

stsp commented on src/base/emu-i386/kvm.c:
```diff
@@ -1021,6 +1045,41 @@ static unsigned int kvm_run(void)
+	  switch(run->mmio.len) {
+	  case 1: write_byte(addr, data[0]); break;
+	  case 2: write_word(addr, *(uint16_t*)data); break;
+	  case 4: write_dword(addr, *(uint32_t*)data); break;
```
kvm seems to allow 8 too.
I suspect we can't hit that, but at least
a "default" catch would be good.
Can some rep movs give 8 btw?

bartoldeman replied:
yes we should have 8, it probably doesn't happen in any code in the wild, but `cmpxchg8b` on vga memory would be problematic for sure.

stsp commented on src/base/emu-i386/kvm.c:
```diff
@@ -1021,6 +1045,41 @@ static unsigned int kvm_run(void)
+	/* read-modify-write instructions give two KVM_EXIT_MMIO
+	   exits in a row before the signal exit */
+      } while (ret == 0 && run->exit_reason == KVM_EXIT_MMIO);
+      assert(ret == -1 && errno == EINTR && run->immediate_exit);
```
Yeah, cool api. :(

stsp commented on src/base/emu-i386/kvm.c:
```diff
@@ -158,6 +158,76 @@ static struct kvm_userspace_memory_region maps[MAXSLOT];
+    sigset_t set;
+    struct kvm_signal_mask *mask;
+
+    kvm_cap_immediate_exit = ioctl(kvmfd, KVM_CHECK_EXTENSION, KVM_CAP_IMMEDIATE_EXIT) > 0;
```
This is likely wrong to do under
`#if !defined(DISABLE_SYSTEM_WA) || !defined(KVM_CAP_IMMEDIATE_EXIT)`
because of the second part of a condition.

stsp replied:
Ah, you re-define it, sorry.

stsp commented on src/base/emu-i386/kvm.c:
```diff
@@ -158,6 +158,76 @@ static struct kvm_userspace_memory_region maps[MAXSLOT];
+
+    /* don't block any signals inside the guest */
+    sigemptyset(&set);
+    mask = malloc(sizeof(*mask) + sizeof(set));
```
Unbelievable piece of crap, why couldn't
they include it in a struct? :(

stsp commented on src/base/emu-i386/kvm.c:
```diff
@@ -158,6 +158,76 @@ static struct kvm_userspace_memory_region maps[MAXSLOT];
+    sigemptyset(&set);
+    mask = malloc(sizeof(*mask) + sizeof(set));
+    mask->len = 8;
+    memcpy(mask->sigset, &set, sizeof(set));
```
I believe it would be better to make
`set` a pointer that points to mask->sigset,
or just directly use `mask->sigset` in all
operations, removing `set`.
Also perhaps `alloca()` is a good candidate
for allocating that discrepancy.

bartoldeman replied:
I could eliminate the dynamical allocation using a struct in the end.
It's an issue between glibc's idea of sigset_t and the kernel's idea, and the struct defintion:
```
/* for KVM_SET_SIGNAL_MASK */
struct kvm_signal_mask {
        __u32 len;
        __u8  sigset[0];
};
```
looking at the kernel code KVM_SET_SIGNAL_MASK errors unless len == 8.
So copying the first 8 bytes of a glibc sigset_t to the sigset above does the trick, as long as at least 7 bytes of padding is added after.

bartoldeman replied:
super-crappy interface for sure... at least it's not used much any more

stsp commented on src/base/emu-i386/kvm.c:
```diff
@@ -158,6 +158,76 @@ static struct kvm_userspace_memory_region maps[MAXSLOT];
+    free(mask);
+  }
+
+  if (kvm_cap_immediate_exit)
```
Duplicate of line 184.

stsp commented on src/base/emu-i386/kvm.c:
```diff
@@ -158,6 +158,76 @@ static struct kvm_userspace_memory_region maps[MAXSLOT];
+  if (kvm_cap_immediate_exit)
+    return;
+
+  if (set)
```
Does this work?

bartoldeman replied:
I tested it but the shadowing is confusing for sure, so using sigset now.

stsp commented on src/base/emu-i386/kvm.c:
```diff
@@ -158,6 +158,76 @@ static struct kvm_userspace_memory_region maps[MAXSLOT];
+{
+  if (kvm_cap_immediate_exit == -1) { // setup
+    struct sigaction sa;
+    sigset_t set;
```
Oops, you got a parameter shadowing here!
I believe this code is not tested, or what am
I missing?

bartoldeman replied:
it was tested, but shadowing is not good in any case, that's for sure!

stsp commented on src/base/emu-i386/kvm.c:
```diff
@@ -1053,6 +1125,7 @@ static unsigned int kvm_run(void)
 	 the kernel with KVM_RUN. The kernel side will first finish
 	 incomplete operations and then check for pending signals." */
       run->immediate_exit = 1;
+      kvm_compat_set_immediate_exit(1);
```
Then at least please move the
`run->immediate_exit = 1;` inside, in the
`#else` block.

bartoldeman replied:
done

stsp commented on src/base/emu-i386/kvm.c:
```diff
@@ -1075,6 +1148,7 @@ static unsigned int kvm_run(void)
       } while (ret == 0 && run->exit_reason == KVM_EXIT_MMIO);
       assert(ret == -1 && errno == EINTR && run->immediate_exit);
       run->immediate_exit = 0;
+      kvm_compat_set_immediate_exit(0);
```
Same, please move `run->immediate_exit = 0;`
inside.

bartoldeman replied:
done

stsp commented:
Thanks!
So what performance improvement is that?

bartoldeman commented:
No performance improvement. Just compatibility with VCPI where we lose control of the page tables, and it'll make some cleanups possible, since KVM_BASE can be seen as physical RAM where paging is effectively disabled.

That is, I think it's now even possible to turn off paging inside KVM for V86, though then you need to turn it on and off when entering and leaving DPMI, so probably a little too annoying.

stsp commented:
But we don't need to decode instruction now?
I thought this can add some.
